--- linux-2.6.21.7/kernel/softlockup.c	2007-08-04 18:11:13.000000000 +0200
+++ linux-2.6.21.7/kernel/softlockup.c	2009-03-10 16:32:07.170160525 +0100
@@ -70,7 +70,7 @@ void softlockup_tick(void)
 		per_cpu(print_timestamp, this_cpu) = touch_timestamp;
 
 		spin_lock(&print_lock);
-		printk(KERN_ERR "BUG: soft lockup detected on CPU#%d!\n",
+		panic(KERN_ERR "BUG: soft lockup detected on CPU#%d!\n",
 			this_cpu);
 		dump_stack();
 		spin_unlock(&print_lock);
