diff -pruN ushare-1.1a.orig/src/cds.c ushare-1.1a.mod/src/cds.c
--- ushare-1.1a.orig/src/cds.c	2008-11-21 14:48:16.000000000 +0100
+++ ushare-1.1a.mod/src/cds.c	2008-11-21 14:47:10.000000000 +0100
@@ -271,7 +271,7 @@ didl_add_param (struct buffer_t *out, ch
 static void
 didl_add_value (struct buffer_t *out, char *param, off_t value)
 {
-  buffer_appendf (out, " %s=\"%lld\"", param, value);
+  buffer_appendf (out, " %s=\"%lld\"", param, (long long)value);
 }
 
 static void
