diff -pNur linux-2.6.21.7/net/ipv4/netfilter/ipt_TOS.c linux-2.6.21.7-patched/net/ipv4/netfilter/ipt_TOS.c
--- linux-2.6.21.7/net/ipv4/netfilter/ipt_TOS.c	2007-08-04 18:11:13.000000000 +0200
+++ linux-2.6.21.7-patched/net/ipv4/netfilter/ipt_TOS.c	2008-06-17 13:03:02.000000000 +0200
@@ -57,8 +57,8 @@ checkentry(const char *tablename,
 	    && tos != IPTOS_RELIABILITY
 	    && tos != IPTOS_MINCOST
 	    && tos != IPTOS_NORMALSVC) {
-		printk(KERN_WARNING "TOS: bad tos value %#x\n", tos);
-		return 0;
+		printk(KERN_WARNING "TOS: bad tos value %#x, accepting\n", tos); /* Efixo mods: accept unnormalized TOS */
+		return 1;
 	}
 	return 1;
  }
