diff -pNur e2fsprogs-1.40.11.orig/lib/blkid/llseek.c e2fsprogs-1.40.11/lib/blkid/llseek.c
--- e2fsprogs-1.40.11.orig/lib/blkid/llseek.c	2008-02-19 05:33:10.000000000 +0100
+++ e2fsprogs-1.40.11/lib/blkid/llseek.c	2008-08-01 18:16:15.000000000 +0200
@@ -60,10 +60,18 @@ extern long long llseek(int fd, long lon
 #ifndef __i386__
 static int _llseek(unsigned int, unsigned long, unsigned long,
 		   blkid_loff_t *, unsigned int);
-
+#if 0
 static _syscall5(int, _llseek, unsigned int, fd, unsigned long, offset_high,
 		 unsigned long, offset_low, blkid_loff_t *, result,
 		 unsigned int, origin)
+#else
+static int _llseek(unsigned int fd, unsigned long offset_high,
+		   unsigned long offset_low, loff_t *result,
+		   unsigned int whence)
+{
+	return 0;
+}
+#endif
 #endif
 
 static blkid_loff_t my_llseek(int fd, blkid_loff_t offset, int origin)
diff -pNur e2fsprogs-1.40.11.orig/lib/ext2fs/llseek.c e2fsprogs-1.40.11/lib/ext2fs/llseek.c
--- e2fsprogs-1.40.11.orig/lib/ext2fs/llseek.c	2008-02-19 05:33:10.000000000 +0100
+++ e2fsprogs-1.40.11/lib/ext2fs/llseek.c	2008-08-01 18:16:21.000000000 +0200
@@ -62,9 +62,18 @@ extern long long llseek (int fd, long lo
 static int _llseek (unsigned int, unsigned long,
 		   unsigned long, ext2_loff_t *, unsigned int);
 
+#if 0
 static _syscall5(int,_llseek,unsigned int,fd,unsigned long,offset_high,
 		 unsigned long, offset_low,ext2_loff_t *,result,
 		 unsigned int, origin)
+#else
+static int _llseek(unsigned int fd, unsigned long offset_high,
+		   unsigned long offset_low, loff_t *result,
+		   unsigned int whence)
+{
+	return 0;
+}
+#endif
 #endif
 
 static ext2_loff_t my_llseek (int fd, ext2_loff_t offset, int origin)
