--- a/minissdp.c
+++ b/minissdp.c
@@ -58,7 +58,15 @@ OpenAndConfSSDPReceiveSocket()
 		syslog(LOG_ERR, "socket(udp): %m");
 		return -1;
 	}	
-	
+
+	i = 1;
+	if(setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &i, sizeof(i)) < 0)
+	{
+		syslog(LOG_ERR, "setsockopt(SO_REUSEADDR): %m");
+		close(s);
+		return -1;
+	}
+
 	memset(&sockname, 0, sizeof(struct sockaddr_in));
     sockname.sin_family = AF_INET;
     sockname.sin_port = htons(SSDP_PORT);
