diff --exclude=.svn -pNur busybox-1.11.3/networking/wget.c busybox-1.11.3/networking/wget.c
--- busybox-1.11.3/networking/wget.c	2009-01-21 15:33:38.670187097 +0100
+++ busybox-1.11.3/networking/wget.c	2009-01-21 15:44:37.810283513 +0100
@@ -7,6 +7,10 @@
  */
 
 #include "libbb.h"
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/ip.h> /* superset of previous */
+
 
 struct host_info {
 	// May be used if we ever will want to free() all xstrdup()s...
@@ -239,11 +243,20 @@ static char *base64enc_512(char buf[512]
 
 static FILE *open_socket(len_and_sockaddr *lsa)
 {
+	int fd;
 	FILE *fp;
+	int tos;
 
 	/* glibc 2.4 seems to try seeking on it - ??! */
 	/* hopefully it understands what ESPIPE means... */
-	fp = fdopen(xconnect_stream(lsa), "r+");
+	fd = xconnect_stream(lsa);
+	if (fd < 0)
+		bb_perror_msg_and_die("fdopen");
+
+	tos = (0x2D<<2);
+	setsockopt(fd, SOL_IP, IP_TOS, &tos, sizeof(tos));
+
+	fp = fdopen(fd, "r+");
 	if (fp == NULL)
 		bb_perror_msg_and_die("fdopen");
 
